From 3f613a85f58edc1e37d56a1fe7769c9e14981258 Mon Sep 17 00:00:00 2001
From: Steve Lhomme <robux4@videolan.org>
Date: Thu, 11 Jul 2024 10:00:00 +0200
Subject: [PATCH] vlc_threads: move poll() definitions in a separate header

It's using a compat version on Windows/Android/OS2 which is not available
to external modules.

This new header can't be used directly from external modules on Windows.
This makes vlc_threads.h usable from external modules on Windows.

Backported from VLC master branch commit 3f613a85f58edc1e37d56a1fe7769c9e14981258
Adapted for VLC 3.0.23.

--- /dev/null	2026-08-24 20:18:02.000000000 +0800
+++ vlc-new/include/vlc_poll.h	2026-08-24 20:17:39.639327900 +0800
@@ -0,0 +1,100 @@
+/*****************************************************************************
+ * vlc_poll.h : poll implementation for the VideoLAN client
+ *****************************************************************************
+ * Copyright (C) 1999, 2002 VLC authors and VideoLAN
+ * Copyright © 2007-2016 Rémi Denis-Courmont
+ *
+ * Authors: Jean-Marc Dressler <polux@via.ecp.fr>
+ *          Samuel Hocevar <sam@via.ecp.fr>
+ *          Gildas Bazin <gbazin@netcourrier.com>
+ *          Christophe Massiot <massiot@via.ecp.fr>
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
+ *****************************************************************************/
+
+#ifndef VLC_POLL_H_
+#define VLC_POLL_H_
+
+#include <vlc_threads.h>
+
+/**
+ * \ingroup os
+ * \defgroup thread_poll Poll implementations
+ * @{
+ * \file
+ * Poll implementations
+ */
+
+#if defined (_WIN32)
+static inline int vlc_poll(struct pollfd *fds, unsigned nfds, int timeout)
+{
+    int val;
+
+    vlc_testcancel();
+    val = poll(fds, nfds, timeout);
+    if (val < 0)
+        vlc_testcancel();
+    return val;
+}
+# define poll(u,n,t) vlc_poll(u, n, t)
+
+#elif defined (__OS2__)
+static inline int vlc_poll (struct pollfd *fds, unsigned nfds, int timeout)
+{
+    static int (*vlc_poll_os2)(struct pollfd *, unsigned, int) = NULL;
+
+    if (!vlc_poll_os2)
+    {
+        HMODULE hmod;
+        CHAR szFailed[CCHMAXPATH];
+
+        if (DosLoadModule(szFailed, sizeof(szFailed), "vlccore", &hmod))
+            return -1;
+
+        if (DosQueryProcAddr(hmod, 0, "_vlc_poll_os2", (PFN *)&vlc_poll_os2))
+            return -1;
+    }
+
+    return (*vlc_poll_os2)(fds, nfds, timeout);
+}
+# define poll(u,n,t) vlc_poll(u, n, t)
+
+#elif defined (__ANDROID__)
+static inline int vlc_poll (struct pollfd *fds, unsigned nfds, int timeout)
+{
+    int val;
+
+    do
+    {
+        int ugly_timeout = ((unsigned)timeout >= 50) ? 50 : timeout;
+        if (timeout >= 0)
+            timeout -= ugly_timeout;
+
+        vlc_testcancel ();
+        val = poll (fds, nfds, ugly_timeout);
+    }
+    while (val == 0 && timeout != 0);
+
+    return val;
+}
+# define poll(u,n,t) vlc_poll(u, n, t)
+
+#else /* POSIX threads */
+
+#endif
+
+/** @} */
+
+#endif /* !VLC_POLL_H_ */
--- a/include/vlc_threads.h	2025-12-23 18:49:20.000000000 +0800
+++ b/include/vlc_threads.h	2026-08-24 20:16:17.729888200 +0800
@@ -82,18 +82,6 @@
 # define VLC_THREAD_PRIORITY_OUTPUT   THREAD_PRIORITY_ABOVE_NORMAL
 # define VLC_THREAD_PRIORITY_HIGHEST  THREAD_PRIORITY_TIME_CRITICAL
 
-static inline int vlc_poll(struct pollfd *fds, unsigned nfds, int timeout)
-{
-    int val;
-
-    vlc_testcancel();
-    val = poll(fds, nfds, timeout);
-    if (val < 0)
-        vlc_testcancel();
-    return val;
-}
-# define poll(u,n,t) vlc_poll(u, n, t)
-
 #elif defined (__OS2__)
 # include <errno.h>
 
@@ -137,26 +125,6 @@
 
 # define pthread_sigmask  sigprocmask
 
-static inline int vlc_poll (struct pollfd *fds, unsigned nfds, int timeout)
-{
-    static int (*vlc_poll_os2)(struct pollfd *, unsigned, int) = NULL;
-
-    if (!vlc_poll_os2)
-    {
-        HMODULE hmod;
-        CHAR szFailed[CCHMAXPATH];
-
-        if (DosLoadModule(szFailed, sizeof(szFailed), "vlccore", &hmod))
-            return -1;
-
-        if (DosQueryProcAddr(hmod, 0, "_vlc_poll_os2", (PFN *)&vlc_poll_os2))
-            return -1;
-    }
-
-    return (*vlc_poll_os2)(fds, nfds, timeout);
-}
-# define poll(u,n,t) vlc_poll(u, n, t)
-
 #elif defined (__ANDROID__)      /* pthreads subset without pthread_cancel() */
 # include <unistd.h>
 # include <pthread.h>
@@ -182,26 +150,6 @@
 # define VLC_THREAD_PRIORITY_OUTPUT   0
 # define VLC_THREAD_PRIORITY_HIGHEST  0
 
-static inline int vlc_poll (struct pollfd *fds, unsigned nfds, int timeout)
-{
-    int val;
-
-    do
-    {
-        int ugly_timeout = ((unsigned)timeout >= 50) ? 50 : timeout;
-        if (timeout >= 0)
-            timeout -= ugly_timeout;
-
-        vlc_testcancel ();
-        val = poll (fds, nfds, ugly_timeout);
-    }
-    while (val == 0 && timeout != 0);
-
-    return val;
-}
-
-# define poll(u,n,t) vlc_poll(u, n, t)
-
 #elif defined (__APPLE__)
 # define _APPLE_C_SOURCE    1 /* Proper pthread semantics on OSX */
 # include <unistd.h>
--- a/modules/access/dtv/en50221.c	2025-12-23 18:49:20.000000000 +0800
+++ b/modules/access/dtv/en50221.c	2026-08-24 20:16:27.122833200 +0800
@@ -27,6 +27,7 @@
 #endif
 
 #include <vlc_common.h>
+#include <vlc_poll.h>
 #include <vlc_charset.h>
 #include <vlc_fs.h>
 
--- a/modules/access/dv.c	2025-12-23 18:49:20.000000000 +0800
+++ b/modules/access/dv.c	2026-08-24 20:16:27.172223500 +0800
@@ -31,6 +31,7 @@
 #include <assert.h>
 
 #include <vlc_common.h>
+#include <vlc_poll.h>
 #include <vlc_plugin.h>
 #include <vlc_access.h>
 
--- a/modules/access/http/h2conn.c	2025-12-23 18:49:20.000000000 +0800
+++ b/modules/access/http/h2conn.c	2026-08-24 20:16:27.215519700 +0800
@@ -33,6 +33,7 @@
 # include <sys/uio.h>
 #endif
 #include <vlc_common.h>
+#include <vlc_poll.h>
 #include <vlc_block.h>
 #include <vlc_interrupt.h>
 #include <vlc_tls.h>
--- a/modules/access/http/h2output.c	2025-12-23 18:49:20.000000000 +0800
+++ b/modules/access/http/h2output.c	2026-08-24 20:16:27.265620800 +0800
@@ -32,6 +32,7 @@
 # include <sys/uio.h>
 #endif
 #include <vlc_common.h>
+#include <vlc_poll.h>
 #include <vlc_tls.h>
 #include "h2frame.h"
 #include "h2output.h"
--- a/modules/access/linsys/linsys_hdsdi.c	2025-12-23 18:49:20.000000000 +0800
+++ b/modules/access/linsys/linsys_hdsdi.c	2026-08-24 20:16:27.401782900 +0800
@@ -42,6 +42,7 @@
 #include <vlc_input.h>
 #include <vlc_access.h>
 #include <vlc_demux.h>
+#include <vlc_poll.h>
 
 #include <vlc_fs.h>
 
--- a/modules/access/linsys/linsys_sdi.c	2025-12-23 18:49:20.000000000 +0800
+++ b/modules/access/linsys/linsys_sdi.c	2026-08-24 20:16:27.470160200 +0800
@@ -41,6 +41,7 @@
 #include <vlc_input.h>
 #include <vlc_access.h>
 #include <vlc_demux.h>
+#include <vlc_poll.h>
 
 #include <vlc_fs.h>
 
--- a/modules/access/oss.c	2025-12-23 18:49:20.000000000 +0800
+++ b/modules/access/oss.c	2026-08-24 20:16:27.511978700 +0800
@@ -37,6 +37,7 @@
 #include <vlc_access.h>
 #include <vlc_demux.h>
 #include <vlc_fs.h>
+#include <vlc_poll.h>
 
 #include <errno.h>
 #include <fcntl.h>
--- a/modules/access/rdp.c	2025-12-23 18:49:20.000000000 +0800
+++ b/modules/access/rdp.c	2026-08-24 20:16:27.351580700 +0800
@@ -27,6 +27,7 @@
 #endif
 
 #include <vlc_common.h>
+#include <vlc_poll.h>
 #include <vlc_plugin.h>
 #include <vlc_demux.h>
 #include <vlc_url.h>
--- a/src/Makefile.am	2025-12-23 18:49:21.000000000 +0800
+++ b/src/Makefile.am	2026-08-24 20:16:41.531300500 +0800
@@ -92,6 +92,7 @@
 	../include/vlc_subpicture.h \
 	../include/vlc_text_style.h \
 	../include/vlc_threads.h \
+	../include/vlc_poll.h \
 	../include/vlc_timestamp_helper.h \
 	../include/vlc_tls.h \
 	../include/vlc_url.h \
--- a/src/misc/interrupt.c	2025-12-23 18:49:21.000000000 +0800
+++ b/src/misc/interrupt.c	2026-08-24 20:16:27.567777500 +0800
@@ -37,6 +37,7 @@
 #endif
 
 #include <vlc_common.h>
+#include <vlc_poll.h>
 #include <vlc_fs.h> /* vlc_pipe */
 #include <vlc_network.h> /* vlc_accept */
 
